agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+            webSocket.closeBlocking()
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+        }
+    }
+
+    override fun connect(url: String, login: String, password: String) {
+        disconnect()
+        webSocket = AgentWebSocketClient(URI(url))
+        webSocket.apply {
+            connectBlocking()
+            sendJson(StateRequest())
+            var stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            if(!checkState(stateResponse, login)) {
+                sendJson(WalletConnect(login, password, id = Instant.now().toEpochMilli().toString()))
+                stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            }
+            if(!checkState(stateResponse, login))
+                throw AgentConnectionException("Error connecting to $url")
+        }
+    }
+
+    private var connectionStatus : AgentConnectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+
+    override fun getConnectionStatus(): AgentConnectionStatus = connectionStatus
+
+    private var counterParty: IndyParty? = null
+
+    private lateinit var webSocket : AgentWebSocketClient
+
+    private fun checkState(stateMessage: State?, userName: String) : Boolean {
+        return if(stateMessage != null &&
+                stateMessage.content?.get("initialized") == true &&
+                stateMessage.content?.get("agent_name") == userName ) {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_CONNECTED
+            true
+        } else {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+            false
+        }
+    }
+
+    override fun acceptInvite(invite: String) {
+        if (invite != null && getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            sendJson(ReceiveInviteMessage(invite))
+            val invite = webSocket.waitForMessageOfType<InviteReceivedMessage>(MESSAGE_TYPES.INVITE_RECEIVED)
+            sendRequest(invite.key)
+            val response = webSocket.waitForMessageOfType<RequestResponseReceivedMessage>(MESSAGE_TYPES.RESPONSE_RECEIVED)
+            counterParty = IndyParty(response.their_did, invite.endpoint)
+        } else {
+            throw AgentConnectionException("AgentConnection object has wrong state")
+        }
+    }
+
+    fun sendJson(obj: Any) = webSocket.sendJson(obj)
+
+    override fun genInvite(): String {
+        webSocket.sendJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE, id = Instant.now().toEpochMilli().toString()))
+        return webSocket.waitForMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).invite
+    }
+
+    override fun waitForInvitedParty() {
+        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            webSocket.waitForMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).also {
```

Comment:
I assume there is non-blocking method

Reply:
Non blocking implementation will be implemented later.

Reply:
backlog + tasks limitation

Reply:
https://github.com/hyperledger-labs/cordentity/projects/1#card-18735696